subsys/usb/class/audio/audio.c

```diff
@@ -0,0 +1,1023 @@
+
+		if (is_entity_valid(audio_dev_data, entity)) {
+			return audio_dev_data;
+		}
+	}
+	return NULL;
+}
+
+/**
+ * @brief Handler for feature unit requests.
+ *
+ * This function handles feature unit specific requests.
+ * If request is properly served 0 is returned. Negative errno
+ * is returned in case of an error. This leads to setting stall on IN EP0.
+ *
+ * @param [in]  dev_data USB audio device data.
+ * @param [in]  pSetup   Information about the executed request.
+ * @param [in]  len      Size of the buffer.
+ * @param [out] data     Buffer containing the request result.
+ *
+ * @return 0 if succesfulf, negative errno otherwise.
+ */
+static int handle_feature_unit_req(struct usb_audio_dev_data_t *audio_dev_data,
+				   struct usb_setup_packet *pSetup,
+				   s32_t *len,
+				   u8_t **data)
+{
+	enum usb_audio_feature_unit_control_selectors control_selector;
+	struct feature_unit_descriptor *fu = NULL;
+	struct usb_audio_feature_unit_evt evt = { .dev = audio_dev_data->common.dev };
+	u8_t ch_num, ch_start, ch_end;
+	static u8_t tmp_data_ptr[3];
+	u8_t data_offset = 0;
+	u8_t fu_id = ((pSetup->wIndex) >> 8) & 0xFF;
+	u8_t device = get_feature_unit(audio_dev_data, &fu, fu_id);
+	int ret = -EINVAL;
+
+	evt.dir = get_fu_dir(fu);
+	ch_num = (pSetup->wValue) & 0xFF;
+	control_selector = ((pSetup->wValue) >> 8) & 0xFF;
+	ch_start = ch_num == 0xFF ? 0 : ch_num;
+	ch_end = ch_num == 0xFF ? get_num_of_channels(fu) : ch_num + 1;
+
+	LOG_DBG("CS: %d, CN: %d, len: %d", control_selector, ch_num, *len);
+
+	/* Error checking */
+	if (!(BIT(control_selector) & (get_controls(fu) << 1UL))) {
+		return -EINVAL;
+	} else if (ch_num >= get_num_of_channels(fu) &&
+		   ch_num != 0xFF) {
+		return -EINVAL;
+	}
+
+	for (int ch = ch_start; ch < ch_end; ch++) {
+		switch (control_selector) {
+		case USB_AUDIO_FU_MUTE_CONTROL:
+			switch (pSetup->bRequest) {
+			case USB_AUDIO_SET_CUR:
+				memcpy(&audio_dev_data->controls[device][ch].mute,
+					(*data + data_offset),
+					sizeof(bool));
+				/* Inform the user APP by callback */
+				if (audio_dev_data->ops->feature_update_cb &&
+				    audio_dev_data->ops) {
+					evt.cs = control_selector;
+					evt.channel = ch;
+					evt.val = &audio_dev_data->controls[device][ch].mute;
+					audio_dev_data->ops->feature_update_cb(evt);
+				}
+				ret = 0;
+				break;
+			case USB_AUDIO_GET_CUR:
+				memcpy(&tmp_data_ptr[data_offset],
+					(u8_t *)&audio_dev_data->controls[device][ch].mute,
+					sizeof(bool));
+				ret = 0;
+				break;
+			default:
+				ret = -EINVAL;
+			}
+			data_offset++;
+			break;
+		case USB_AUDIO_FU_VOLUME_CONTROL:
+		case USB_AUDIO_FU_BASS_CONTROL:
+		case USB_AUDIO_FU_MID_CONTROL:
+		case USB_AUDIO_FU_TREBLE_CONTROL:
+		case USB_AUDIO_FU_GRAPHIC_EQUALIZER_CONTROL:
+		case USB_AUDIO_FU_AUTOMATIC_GAIN_CONTROL:
+		case USB_AUDIO_FU_DELAY_CONTROL:
+		case USB_AUDIO_FU_BASS_BOOST_CONTROL:
+		case USB_AUDIO_FU_LOUDNESS_CONTROL:
+			break;
```

Comment:
I left this for the future as by this PR I am only supporting mute feature and nothing more. I wanted to underline there that additional support could be implemented. Mute is core feature that Host requires for the  device to work.

It could return last read mute values only if user chose property in the DT. It's explicitly said there that those are currently not supported.
To avoid this I instead of `break;` i can put `return -ENOTSUP`

Reply:
Removed unused cases.